### PR TITLE
Issue #340 Large event support

### LINKED ISSUE #340: Implement Large Event Support
**Problem description**
For large events we need to add additional support

**Problem location**
writer/event_writer.rs

**Suggestions for an improvement**
Add a flush implementation and add use transient segments to handle large events

### PR DESCRIPTION
**Pull request by vangork**
**Change log description**  
Add Large event Support to Rust Client

**Purpose of the change**  
Fixes #340 

**What the code does** 
Use transient segments to handle large events with size over 8MB, 

**How to verify it**  
```
cd examples
cargo run --bin event_write_and_read
```

### CODE REVIEW

File: src/segment/reactor.rs [hunk 1]
```diff
@@ -49,6 +50,15 @@ impl Reactor {
...
         match event {
             Incoming::AppendEvent(pending_event) => {
+                if pending_event.data.len() > PendingEvent::MAX_WRITE_SIZE {
+                    let mut large_event_writer =
+                        LargeEventWriter::new(selector.delegation_token_provider.clone());
+                    large_event_writer
+                        .write(factory, selector, pending_event)
+                        .await
+                        .map_err(|_| "large event writer failed to write")?;
+                    return Ok(());
```
Comment (tkaitchuck) [inline src/segment/reactor.rs:60]: To preserve order we need to make sure to first flush the prior events.

File: wire_protocol/src/commands.rs [hunk 2]
```diff
@@ -28,7 +28,7 @@ pub const OLDEST_COMPATIBLE_VERSION: i32 = 5;
 pub const TYPE_SIZE: u32 = 4;
 pub const TYPE_PLUS_LENGTH_SIZE: u32 = 8;
 pub const MAX_WIRECOMMAND_SIZE: u32 = 0x00FF_FFFF; // 16MB-1
-
```
Comment (tkaitchuck) [inline wire_protocol/src/commands.rs:31]: The wire_version  field needs to be updated to reflect which wire-commands are being included. (Check to be sure what changes have happened between 11 and 15 in Java and make any updates here.)

File: src/event/writer.rs [hunk 3]
```diff
@@ -82,8 +82,8 @@ pub struct EventWriter {
 
 impl EventWriter {
     pub(crate) const MAX_EVENT_SIZE: usize = 8 * 1024 * 1024;
-    // maximum 16 MB total size of events could be held in memory
-    const CHANNEL_CAPACITY: usize = 16 * 1024 * 1024;
+    // maximum 1GB + 8B total size of events could be held in memory
+    const CHANNEL_CAPACITY: usize = 1024 * 1024 * 1024 + 8;
```
[the remark this answers is not in the document]
Comment (vangork) [reply src/event/writer.rs]: reverted back.

[the remark this answers is not in the document]
Comment (vangork) [reply src/event/writer.rs]: Done

File: src/segment/large_writer.rs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+                .await
+        {
+            Ok(uri) => uri,
+            Err(e) => return Err(LargeEventWriterError::RetryControllerWriting { err: e }),
+        };
+        current_span().record("host", &field::debug(&uri));
+
+        let raw_client = factory.create_raw_client_for_endpoint(uri);
+
+        // create transient segment
```
[the remark this answers is not in the document]
Comment (vangork) [reply src/segment/large_writer.rs]: Done

File: src/segment/large_writer.rs [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+            _ => {
+                info!("creating transient segment failed due to {:?}", reply);
+                return Err(LargeEventWriterError::WrongReply {
+                    expected: String::from("SegmentCreated"),
+                    actual: reply,
+                });
+            }
+        };
+
+        // setup append
```
[the remark this answers is not in the document]
Comment (vangork) [reply src/segment/large_writer.rs]: Done

File: src/segment/large_writer.rs [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                    actual: reply,
+                });
+            }
+        };
+        let mut expected_offset: i64 = 0;
+        let chunks = event.data.chunks(EventWriter::MAX_EVENT_SIZE);
+        for (event_number, chunk) in (0_i64..).zip(chunks) {
+            let data = chunk.to_vec();
+
+            // send ConditionalBlockEnd
```
[the remark this answers is not in the document]
Comment (vangork) [reply src/segment/large_writer.rs]: Done

File: src/segment/large_writer.rs [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+                    return Err(LargeEventWriterError::WrongReply {
+                        expected: String::from("DataAppended"),
+                        actual: reply,
+                    });
+                }
+            };
+            expected_offset += EventWriter::MAX_EVENT_SIZE as i64;
+        }
+
+        // merge segments
```
[the remark this answers is not in the document]
Comment (vangork) [reply src/segment/large_writer.rs]: Done

File: wire_protocol/src/commands.rs (hunk 2 above)
Comment (vangork) [reply wire_protocol/src/commands.rs:31]: CreateTransientSegment is added under `WIRE_VERSION` of 14. But rust client is missing `GetTableSegmentInfo` under `WIRE_VERSION` of 12 and `FlushToStorage` under `WIRE_VERSION` of 12. How to proceed?

Comment (vangork) [reply wire_protocol/src/commands.rs:31]: Done

File: src/segment/reactor.rs (hunk 1 above)
Comment (tkaitchuck) [reply src/segment/reactor.rs:60]: As that is in a caller we can't be sure it was actually done at this point in the code. So here there should either be an assertion to check that it was done by the caller, or it should be done here.

Comment (vangork) [reply src/segment/reactor.rs:60]: `self.flush().await` can make sure that all the pior events has been acked and then send the large event to the reactor task to proceed. In the reactor task, do you think we should check the pending queue size again? if it is larger then zero, make an assert?

Comment (tkaitchuck) [reply src/segment/reactor.rs:60]: Yes. That sounds good.

Comment (vangork) [reply src/segment/reactor.rs:69]: done

**Review by tkaitchuck — APPROVED** [review]

**Comment by tkaitchuck** [comment]
@vangork Test is failing. Please fix.

**Comment by vangork** [comment]
> @vangork Test is failing. Please fix.

Fixed.

**Comment by tkaitchuck** [comment]
Not sure why, but a lot of node/python tests are failing with this branch. We need to resolve this before merging.

**Review by tkaitchuck — APPROVED** [review]

**Pull request merged**